### PR TITLE
Use same max_age regardless of leader/not-leader

### LINKED ISSUE #9357: Vote hash mismatch on mainnet-beta/validator-us-east1-b
`validator-us-east1-b` when delinquent last night after slot 4187115.    At slot 4187144, `validator-us-east1-b` produced a different bank hash than `validator-us-west1-b` and seemingly the rest of the cluster

validator-us-east1-b:
```
[2020-04-07T08:31:22.809585778Z INFO  solana_runtime::bank]  frozen: 4187144
  hash: FmRsY4WgARgNXf1dW9dXzGXWPvKNp4LoKZYovUSPWMKU 
  accounts_delta: Uu2n1bkKHdv812mYR7KEgfqptoiVJM9CKagE1QqnwnF
  signature_count: 190
  last_blockhash: jeKFjTNiWbkpU2u9KmaxGBM8rTnqGTCddnDTtxd4o9A
```
Larger `validator-us-east1-b` log snippet at https://pastebin.com/md2VWphi

validator-us-west1-b:
```
[2020-04-07T08:31:23.082385197Z INFO  solana_runtime::bank] bank frozen: 4187144
  hash: 5M6n9bmiqdrRjr8myHMa2LWJqLhHLEs3WyMGfFniCGpH 
  accounts_delta: CqxeRSyM5dPBKjySz2DEnGM93Rdw6uoTDPYPpJYMWC5S
  signature_count: 190
  last_blockhash: jeKFjTNiWbkpU2u9KmaxGBM8rTnqGTCddnDTtxd4o9A
```

accounts_delta is different.

Full log for both validators at https://drive.google.com/drive/folders/19448bVE3TFRe4rFciKMb6SB9nkOeHw0l?usp=sharing

### PR DESCRIPTION
**Pull request by ryoqun**
(I'm really tired at this time...)

## Problem

For deterministic execution of nonse TX error codepath, we must use consistent max_age for past TX expiration.

#### Leader:

`Bank::load_and_execute_transactions()` is called with **`MAX_PROCESSING_AGE`** from `BankingStage::process_and_record_transactions_lock()`

#### Not-leader/ledger-tool (=blockstore_processor):

`Bank::load_and_execute_transactions()` is called with **`MAX_RECENT_BLOCKHASHES`** from `blockstore_processor::execute_batch()`

And in some corner case, this causes this:

```
if the given identical nonce tx causes insufficient fund error....
    Leader:     It updates nonce account and stores the pay destination account (correct)
             (but rejected by supermajority of other validators)
    Non-leader: It don't update nonce account and don't store the pay destination account. (wrong)

(In both cases, fees are collected as a side-effect, universally.)
```

because of this observed difference of hash age result:

#### leader (nonce is updated):

```
[2020-04-10T09:37:23.453277367Z ERROR solana_runtime::accounts] collect_accounts: Err(InstructionError(1, CustomError(1))) Some(DurableNonce(xdpCWLrjxTb8rNFTAco4E9jt2awraosXKtU2kCVxb92, Account { lamports: 1123456789 data.len: 80 owner: 11111111111111111111111111111111 executable: false rent_epoch: 5 data: 00000000010000005c48a3b41fe5a71522ca2d65ac355b1fc253107096c89e240bc3dcca71b9efb8b46e4f4a3e5dfba8473a7b581bb83b703f2c63c4a3b4c385 hash: FRihkb3Jv6V1DtyWcnqxqtoufiBCxwjXD8x8rVVjrTHQ }))

```

#### not leader (nonce is NOT updated):
```
[2020-04-10T09:37:23.514004924Z ERROR solana_runtime::accounts] collect_accounts: Err(InstructionError(1, CustomError(1))) Some(Extant)
```

```patch
     fn collect_accounts_to_store<'a>(
         &self,
         txs: &'a [Transaction],
         txs_iteration_order: Option<&'a [usize]>,
         res: &'a [TransactionProcessResult],
         loaded: &'a mut [(Result<TransactionLoadResult>, Option<HashAgeKind>)],
         rent_collector: &RentCollector,
         last_blockhash: &Hash,
     ) -> Vec<(&'a Pubkey, &'a Account)> {
+        error!("collect_accounts: {} txs", txs.len());
+
         let mut accounts = Vec::with_capacity(loaded.len());
         for (i, ((raccs, _hash_age_kind), tx)) in loaded
             .iter_mut()
             .zip(OrderedIterator::new(txs, txs_iteration_order))
             .enumerate()
         {
             if raccs.is_err() {
                 continue;
             }
             let (res, hash_age_kind) = &res[i];
+            error!("collect_accounts: {:?} {:?}", res, hash_age_kind);
             let maybe_nonce = match (res, hash_age_kind) {
                 (Ok(_), Some(HashAgeKind::DurableNonce(pubkey, acc))) => Some((pubkey, acc)),
                 (
                     Err(TransactionError::InstructionError(_, _)),
                     Some(HashAgeKind::DurableNonce(pubkey, acc)),
                 ) => Some((pubkey, acc)), // <= SHOULD REACH HERE!!!
                 (Ok(_), _hash_age_kind) => None,
                 (Err(_), _hash_age_kind) => continue,    // <= WRONG REACHED HERE!!!
```

And causing these missing nonce account updates for the validators but the leader:

```patch
 SysvarS1otHistory11111111111111111111111111:
   - lamports: 1
   - owner: 'Sysvar1111111111111111111111111111111111111'
   - executable: false
   - data: '4rmdYzwEyfJhmVo3p18YKQUsnrfh6NpJdTnebJgpg3KoxEsTLryKVtZ2kjhb8MbdmSdoqL4YcXuT36fKB443dnxtJb3wHqNWeqoXNDLHddTcbF8982TABTHwn6UvYmkhbJmqo4vJ189ZTwtsSf9Kt5478BCYuiqzB5cQkMw4iA4x1BvFJVnYDRMhu944KbEpAAFUamz2K5dFkbx6oSSXHZauaenHQB5SiZZHad86r7mQLmbPpSPq3QGSLm8FXDSkvM7jtAjrNnu5HRwN4zaBKUNRZRwkFrX1dFJJHXdBSWh>
quGeST64yhxsL58xPrvzm2uUwJUDf6D7edwkZijBTgoTG7DNSvCGH1VdUSx4p5A4sEhts7KoVL3Lu32uTwtS5FRBc8B9D13WnM6oFuRtuXpnvV8pTWRq7YCQtwTE5qgbNVQNq3jbAT64Z4jtP55dW9oF4sCVB7aHi1N2ZmaprQyEQUhca5SxmGWQzb6Fa8i5spipgkPhHe5mQ8tyexvQSqkNNbBDuTJnoRdjPzDnAuV6h1rCEnzouMDvyLkpzGzuxfoWkvacxQJBzhodTCe5B8E5N7MDLucVz3ZtbeNFJ9HDRZUdjPXGMqi>
   - data_len: 131097
-xdpCWLrjxTb8rNFTAco4E9jt2awraosXKtU2kCVxb92:
-  - lamports: 15123456789
-  - owner: '11111111111111111111111111111111'
-  - executable: false
-  - data: '11116bv5oki4iftAgi9rancYuWZ3qnKyaLjaP75dLXwUvFRk4CMh7q4aWeLgq1yWW1H2cn96Z2fRUvf1qbCLTHVd6ozarwHoniy5j1FnTJP'
-  - data_len: 80
 58xXjpcSGtEtBo3SfousER3SDUfAcUyG6PBYyxVB587h:
   - lamports: 499963177500
   - owner: '11111111111111111111111111111111'
   - executable: false
   - data: ''
   - data_len: 0
```

Resulting these bank hash input difference:

#### leader

```
Apr  7 08:31:22 validator-us-east1-b validator.sh[8909]: [2020-04-07T08:31:22.809585778Z INFO  solana_runtime::bank] bank frozen: 4187144 hash: FmRsY4WgARgNXf1dW9dXzGXWPvKNp4LoKZYovUSPWMKU accounts_delta: Uu2n1bkKHdv812mYR7KEgfqptoiVJM9CKagE1QqnwnF signature_count: 190 last_blockhash: jeKFjTNiWbkpU2u9KmaxGBM8rTnqGTCddnDTtxd4o9A
Apr  7 08:31:22 validator-us-east1-b validator.sh[8909]: [2020-04-07T08:31:22.809604179Z INFO  solana_runtime::bank] accounts hash slot: 4187144 stats: BankHashStats { num_removed_accounts: 356, num_added_accounts: 0, num_lamports_stored: 91925590448565, total_data_len: 747219, num_executable_accounts: 0 }

```

#### not leader/ledger-tool

```
Apr  7 08:31:23 validator-us-west1-b validator.sh[8135]: [2020-04-07T08:31:23.082385197Z INFO  solana_runtime::bank] bank frozen: 4187144 hash: 5M6n9bmiqdrRjr8myHMa2LWJqLhHLEs3WyMGfFniCGpH accounts_delta: CqxeRSyM5dPBKjySz2DEnGM93Rdw6uoTDPYPpJYMWC5S signature_count: 190 last_blockhash: jeKFjTNiWbkpU2u9KmaxGBM8rTnqGTCddnDTtxd4o9A
Apr  7 08:31:23 validator-us-west1-b validator.sh[8135]: [2020-04-07T08:31:23.082403977Z INFO  solana_runtime::bank] accounts hash slot: 4187144 stats: BankHashStats { num_removed_accounts: 354, num_added_accounts: 0, num_lamports_stored: 91835975463565, total_data_len: 747139, num_executable_accounts: 0 }
```

_**And inevitable bank_hash mismatch, forcing the leader at the time of processing out of circle by vote hash mismatch, ultimately**_

## Summary of Changes

I confirmed after back-porting this to v1.0, the vote mismatches don't happen.

Also, strictly speaking, this changes the execution semantics; so a variant of hard-fork.

- ~[ ] todo: write a test?~ (I'd skip)
- [x] todo: audit other callsites?

Fixes #9357

## Random

This was the 2nd hardest bug to fix in solana, with #7736 being the most. ;)

### CODE REVIEW

**Comment by sakridge** [comment]
dang, good find!

**Comment by ryoqun** [comment]
@sakridge @t-nelson Thanks!

@mvines I think this should be released to the mainnet-beta rather quickly. Yet, this should be done rather at once if done as-is. That's because as validators roll up this change, it slowly increases the minitory behavior. Currently, only the leader do the right behavior, so it's ignored. But say 30 validator updated, and if a problematic nonce tx is arrived at the cluster, the fork because of in-transition nature could have a sizable impact...

Or we should add some simple condition slot > 1234 so that this should not immediately take effect.

Sadly, I'm really out of my stamina and time (Friday midnight, here). First, I'll take some sleep. :)

Good night!

**Comment by t-nelson** [comment]
`RecentBlockhashes` sysvar prevents us from changing the leader behavior instead (which would be instantly compatible).  It stores `MAX_PROCESSING_AGE` blockhashes and `FeeCalculators`, so changing it is an ABI break :confused:

**Comment by mvines** [comment]
> Or we should add some simple condition slot > 1234 so that this should not immediately take effect.

We can use squirt some code into
https://github.com/solana-labs/solana/blob/master/genesis-programs/src/lib.rs#L116-L129
on the v1.0 branch.  At "current epoch + 1", we switch over if we're running under `OperatingMode::Stable`.   Then I can coordinate the upgrade with the mainnet-beta validators, "you have ~2 days to upgrade".   For TdS, we can do the same on the v1.1 branch  if we're running under `OperatingMode::Preview`.    This'll probably require a non-serde field in Bank that tells it when to switch over to using MAX_PROCESSING_AGE
 
> Sadly, I'm really out of my stamina and time (Friday midnight, here). First, I'll take some sleep. :)
 
You rock, thanks!

**Comment by mvines** [comment]
Actually better plan: `master`/`v1.1` -- don't care, make the change and get TdS to upgrade without the safety net.   Worse case we get to practice another restart.

on v1.0, `execute_batch()` checks for `OperatingMode::Stable` and uses the old or new value depending on the epoch, so we give mainnet-beta validators an epoch to upgrade.

**Comment by solana-grimes** [comment]
:broken_heart: Unable to automerge due to CI failure

File: ledger/src/blockstore_processor.rs [hunk 1]
```diff
@@ -70,7 +70,7 @@ fn execute_batch(
         balances,
     ) = batch.bank().load_execute_and_commit_transactions(
         batch,
-        MAX_RECENT_BLOCKHASHES,
```
Comment (ryoqun) [inline ledger/src/blockstore_processor.rs:73]: my quick git time-travel told that this bug is introduced sooo long ago: #1499 201**8** Oct PR.

**Comment by garious** [comment]
@ryoqun, @sakridge, @mvines, @t-nelson  I would think that this change would cause the Devnet and Testnet clusters to stall regularly, as the leader would occasionally include transactions that were just about to expire. Nearby validators might accept them and distant validators would reject them, causing a fork.  It's easy to imagine this causing something near a 50/50 fork, which take the longest to resolve.

**Comment by ryoqun** [comment]
@garious Thanks for heads-up!
Well, I don't think there is problem with this pr. My github description sucks to explain but this pr aligns with `MAX_PROCESSING_AGE` for all cases. Previously, there are case validators used `MAX_PROCESSING_AGE` (which is `MAX_RECENT_BLOCKHASHES/2`) or `MAX_RECENT_BLOCKHASHES` depending on the codepath.

Or I might be missing. Could you explain the difference of situation between nearby validators and distant validators a bit?

**Comment by garious** [comment]
I think you're missing. In the original code, the leader rejects any transaction that isn't within the last 150 most recent blockhashes, such that it can send a block to all the validators within the next 150 blockhashes. As long as the validators validate the transaction within 300 blockhashes from when the client received the most recent blockhash, the transaction will be accepted. If the leaders and validators reject transactions at the same blockhash and the leader includes a transaction with the oldest blockhash, then the validator will surely purge the oldest blockhash before the leader's block traverses the network, causing the block to be rejected.

**Comment by ryoqun** [comment]
@garious Thanks for further explanation. I now understands your concern well.

Still, I think that problem doesn't exist. That's because I think a particular bank at slot XXX handles expiration (blockhash check) deterministically with the identical subview into the blockhashes with same depth (XXX - 150) regardless being leader (producing a block) / distant validaotor (replaying the block). So, even if a validator processes a block much later since produced, it always see the same set of blockhashes (also, statuscaches, for that matter). Their subviews are strictly associated with each bank and they're strictly updated at each slot boundary as sysvars.

My guess is that we lost the justification of distinction between `MAX_RECENT_BLOCKHASHES` and `MAX_PROCESSING_AGE` via a course of introduction of nonces and general further determinism? (CC: @t-nelson )

I've glanced over the relevant code once. But I couldn't spot suspicious code where delayed validator 
could wrongly purge the blockhash queue prematurely and naively process transactions after that.

Also, this patch landed 1 month ago. If there were actually consensus bug, it should have happened already, considering tds is considerably latency-heavy compared to the mainnet-beta.

Lastly, well I should have written an accompanying unit-test for this. This discussion is enough to justify to do so, I guess. :)

**Comment by ryoqun** [comment]
> EDIT: Well, I now understand I was wrong, considering the actual wildlife observation... Please disregard my following comment:
> 
> > @garious Thanks for further explanation. I now understands your concern well.
> > Still, I think that problem doesn't exist. That's because I think a particular bank at slot XXX handles expiration (blockhash check) deterministically with the identical subview into the blockhashes with same depth (XXX - 150) regardless being leader (producing a block) / distant validaotor (replaying the block). So, even if a validator processes a block much later since produced, it always see the same set of blockhashes (also, statuscaches, for that matter). Their subviews are strictly associated with each bank and they're strictly updated at each slot boundary as sysvars.
> > My guess is that we lost the justification of distinction between `MAX_RECENT_BLOCKHASHES` and `MAX_PROCESSING_AGE` via a course of introduction of nonces and general further determinism? (CC: @t-nelson )
> > I've glanced over the relevant code once. But I couldn't spot suspicious code where delayed validator
> > could wrongly purge the blockhash queue prematurely and naively process transactions after that.
> > Also, this patch landed 1 month ago. If there were actually consensus bug, it should have happened already, considering tds is considerably latency-heavy compared to the mainnet-beta.
> > Lastly, well I should have written an accompanying unit-test for this. This discussion is enough to justify to do so, I guess. :)

Well, this turned out to be CORRECT. Such an unexpected ending. I'll document and write tests about these findings. :)

**Comment by garious** [comment]
My mistake. Sorry for sending you on a wild goose chase.

**Comment by ryoqun** [comment]
@garious no problem! I always welcome heads-up. And I think the relevant code is more scrutinized now. :)

**Pull request merged**